libs/core/execution_base/include/hpx/execution_base/traits/coroutine_traits.hpp

```diff
@@ -0,0 +1,231 @@
+    // context it will generally need to type-erase the
+    // coroutine_handle<Promise> to coroutine_handle<void> so that it can
+    // store the continuation for an arbitrary coroutine-type.
+    // If theawait_suspend() method overload-set only has overloads that
+    // accept specific types of coroutine_handle<P> then it is only awaitable
+    // within specific contexts and thus we don’t consider it to satisfy the
+    // Awaiter concept.
+
+    template <typename Awaiter, typename Promise = void>
+    struct is_awaiter
+      : detail::is_awaiter_impl<detail::has_await_ready<Awaiter> &&
+                detail::has_await_resume<Awaiter> &&
+                (detail::has_await_suspend<Awaiter> ||
+                    detail::has_await_suspend_coro_handle<Awaiter, Promise>),
+            Awaiter, Promise>
+    {
+    };
+
+    template <typename Awaiter, typename Promise = void>
+    bool inline constexpr is_awaiter_v = is_awaiter<Awaiter, Promise>::value;
+
+    namespace detail {
+
+        template <typename Awaitable, typename = void>
+        constexpr bool has_member_operator_co_await_v = false;
+
+        template <typename Awaitable>
+        constexpr bool has_member_operator_co_await_v<Awaitable,
+            std::void_t<decltype(
+                std::declval<Awaitable>().operator co_await())>> = true;
+
+        template <typename Awaitable, typename = void>
+        constexpr bool has_free_operator_co_await_v = false;
+
+        template <typename Awaitable>
+        constexpr bool has_free_operator_co_await_v<Awaitable,
+            std::void_t<decltype(operator co_await(
+                std::declval<Awaitable>()))>> = true;
+
+        template <typename, typename = void, typename...>
+        constexpr bool has_await_transform{};
+
+        template <typename Promise, typename... Ts>
+        constexpr bool has_await_transform<Promise,
+            std::void_t<decltype(std::declval<Promise>().await_transform(
+                ((Ts &&) std::declval<Ts>())...))>,
+            Ts...> = true;
+
+    }    // namespace detail
+
+    // Returns the result of applying operator co_await() to the function’s
+    // argument, if the operator is defined, otherwise returns a reference
+    // to the input argument.
+    template <typename Awaitable>
+    decltype(auto) get_awaiter(Awaitable&& await, void*)
+    {
+        if constexpr (detail::has_member_operator_co_await_v<Awaitable>)
+        {
+            return ((Awaitable &&) await).operator co_await();
+        }
+        else if constexpr (detail::has_free_operator_co_await_v<Awaitable>)
+        {
+            return operator co_await((Awaitable &&) await);
+        }
+        else
+        {
+            return (Awaitable &&) await;
+        }
+    }
+
+    template <typename Awaitable, typename Promise,
+        typename = std::enable_if_t<
+            detail::has_await_transform<Promise, Awaitable>, void>>
```

Comment:
the `void` is not needed (it's the default for `enable_if`)

Reply:
Noted